2 halo term for CCL and NC backends

combet opened the pull request:
This implements the surface and excess surface density from the 2-h term for CCL and NC, using eq. (13) from[ Oguri& Hamana (2011)](https://ui.adsabs.harvard.edu/abs/2011MNRAS.414.1851O/abstract)

- The matter power spectrum P(k) has a benchmark test
-  The default (and only option so far) is to have sigma8=0.8 (because of the way CLMM cosmology objects are currently set up.). I think we'll need to implement sigma8 as a user-defined parameter. Straightforward for CCL as it is a user-defined parameter there, but less so for NumCosmo as it requires normalizing the primordial PS amplitude to match a given value of sigma8. Do it in a new issue, or here?
- We have no benchmark for the 2-h term yet. We can probably make this another specific issue. For the 2h-term, the only tests available so far check:
     - that we do get a "NotImplementedError" if the backend is CT.
     - that it "runs" when CCL or NC is the backend; 
- The demo_theory notebooks (both OO and functional interface) have been updated to call the new functions
- The halo bias is only included as a `float` keyword to be informed by the user. A next step would be to implement a b(M) model, but that can come later I think

pennalima commented:
Hi @combet, CT computes the 2-halo term. Right? 
Should we discuss this in another issue?

combet commented:
Hi @pennalima - CT has a 2 halo term options but takes as input the matter power spectrum as it has no functionality to compute it itself. So I don't think we should invest time to get the 2-h term for CT as the user will first need to compute the PS using external software and pass the PS as an argument to CT.

combet commented:
Hi @m-aguena - I was updating the implementation of the 2h term, following the initial work that @calumhrmurray did during the sprint week. But I just noticed all the commits coming from Shenming's work to use CLMM on real data in that branch. There must have been some mix up as, if I recall correctly, Calum was simultaneously looking at Shenming's work at that time. I'm wondering what is the best course of action:

1. I believe all of Shenming's commits only have to do with the new `real_data` notebook in the example folder. So removing that file should be sufficient and is straightforward. The log history would still have the all the unwanted commits there, though.
2. Instead should we try to revert all of the unwanted commits, to clean up the history?

m-aguena commented:
@combet I noticed there was a wrong merge of branch issue/448 into this one at some point:

```Merge branch 'issue/448/read_data' of https://github.com/LSSTDESC/CLMM into issue/446/ccl_2_halo_term```

but I don't think this will be a major problem. We can just remove the nb in this branch, and remove the related comments when squashing and merging this PR.

Just leaving a reminder here in case this causes any conflicts with PR #450 .

combet commented:
@m-aguena - thanks! Yes of course, I forgot that squash and merge would allow to remove all the unwanted comments. I'll proceed with the first option then and will mark the PR ready for review afterwards.

hsinfan1996 commented:
There is a .DS_Store file, which should be removed.

m-aguena commented:
@combet I moved the 2 halo term to the parent class, now it should work with any backend that provides a power spectrum.

@pennalima can you fill the functions `NumCosmoCosmology._get_rho_m` and `NumCosmoCosmology._get_rho_m`? Then we can test the 2 halo term with numcosmo.

combet commented on clmm/cosmology/ccl.py:
```diff
@@ -73,12 +73,17 @@ def _get_Omega_m(self, z):
         return ccl.omega_x(self.be_cosmo, 1.0/(1.0+z), "matter")
```
Done

combet commented:
@pennalima @eduardojsbarroso - If you get the chance, could you please check:
- the implementation of the NC linear matter power spectrum I added to `cosmology/numcosmo.py`.The function is called `_eval_linear_matter_powerspectrum()`. As we have BBKS as the default transfer function for the CCL backend, I used that here as well. Also, I adjusted the value of As to make sure we have the same value as the default CCL cosmology. Not sure if this is the best way to proceed.

- the implementation of the `_get_rho_m()` function, in the same file. The goal is to have the physical matter density in Msun/Mpc3, which is required for the 2h term. Maybe there's a more direct way to obtain it from NC, but I haven't found it :)

eduardojsbarroso commented:
@combet I wasn't able to find a more direct way to implement the ´_get_rho_m()´ function. The only other function that I found that was able to compute the ´OMEGA_M´ is in the LCDM Comology object from NumCosmo, and returns the same as the function that you called (´self.be_cosmo.E2Omega_m(z)´). 

But maybe instead of calling 
´self._get_param('h') * self._get_param('h')´
You could call
´self.be_cosmo.crit_density ()´  (This is the function ´nc_hicosmo_crit_density ()´ on the NumCosmo library).

For the other function, I don't understand fully how to compute the spectrum to help you.

combet commented:
Thanks @eduardojsbarroso for the rho_m confirmation.

Following up on the P(k) question, I was using the BBKS transfer function. Switching to the Eisenstein & Hu TF for the CCL and NC backend give nice aggrement for the 2-h term (see relative difference figure below)
<img width="422" alt="Screenshot 2021-12-10 at 15 44 21" src="https://user-images.githubusercontent.com/20227809/145592105-30c91396-b28e-4b18-aa1c-a07e1f22f1eb.png">

So I switch to using EH by default for the computation of the 2-h term.

m-aguena commented:
I updated some tests, but we still need to put some actuall values to test the power spectrum and the 2h term

combet commented:
@m-aguena, I added a benchmark test for P(k). 

When implementing the 2h term, we added the `_get_rho_m` functions to both the `ccl` and `numcosmo` backends of the Cosmology class, using builtin functions from ccl and numcosmo. But I see that it is also defined in a generic manner the `parent_class`. Should we move this generic implementation to the `cluster_toolkit` backend instead?

combet commented:
@calumhrmurray has implemented the 2h term for the surface density (before we only had the excess surface density) and I have updated the notebooks and tests accordingly. This is again "ready for review" @m-aguena, @marina-ricci

m-aguena commented:
PR converted to draft as per the new procedure (labels `PR: Ready for Review` and `PR: Not ready` will not be used anymore). Please use the button `Ready for review` if/when this PR is ready.

combet commented:
Following the discussion at the last tag up, I've (re-)added the option to have a `halobias` in the 2-halo term functions. At the moment a `float` is requested, i.e, the halobias must be computed beforehand by the user, e.g. using CCL.

pennalima reviewed (approved):
Iẗ would be better to define A_s parameter (power spectrum normalization) instead of sigma_8. The last is a derived parameter. However, CCL just computes the "analytic" power spectrum if the CCL cosmology is instantiated with sigma8 (it doesn´t work with A_s). So I didn't modify the code.

combet commented:
Thanks @pennalima - so I think I'll open another issue to find the best way to deal with sigma8/As in the CLMM cosmology. At the moment, if generating the cosmology object directly from CLMM (and not using a per-instantiated cosmology from the backedn), the 2-halo term will use sigma8=0.8. Would be good find a way to add flexibility within the CLMM cosmology.

combet commented:
@pennalima, yes that's a good point. I'll merge the PR.
